src/platform_impl/windows/ime.rs

```diff
@@ -31,21 +31,32 @@ impl ImeContext {
         let text = unsafe { self.get_composition_string(GCS_COMPSTR) }?;
         let attrs = unsafe { self.get_composition_data(GCS_COMPATTR) }.unwrap_or_default();
 
+        // text is UTF-8, and attrs is UTF-16, so we need to convert the UTF-8 index to UTF-16 for
+        // processing.
+        let utf16_to_utf8 = {
+            let mut utf16_to_utf8 = Vec::new();
+            let mut utf8_offset = 0;
+
+            for chr in text.chars() {
+                let utf16_len = chr.len_utf16();
+                utf16_to_utf8.extend(std::iter::repeat(utf8_offset).take(utf16_len));
+                utf8_offset += chr.len_utf8();
+            }
+            utf16_to_utf8
+        };
+
         let mut first = None;
         let mut last = None;
-        let mut boundary_before_char = 0;
 
-        for (attr, chr) in attrs.into_iter().zip(text.chars()) {
+        for (utf16_idx, &attr) in attrs.iter().enumerate() {
```

Comment:
Is it possible to iterate `text.chars()` and accumulate the `utf16_len` and `utf8_len` at once, the nuse `utf16_len` to offset into the `attrs` array and check whether we should start a boundary which we accumulate with the `utf8_len`? So we avoid `utf16_to_utf8` map, since we'll be doing it on the fly.

Reply:
@kchibisov 
Thanks for your review.
The difference from the original code is smaller and easier to read when calculated on the fly. This fix worked as expected for me.

https://github.com/rust-windowing/winit/pull/4201/commits/5aa687f1da5853c784e627194fa7b3a2869d3c97